src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+import Debug "mo:base/Debug";
+import Error "mo:base/Error";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Principal "mo:base/Principal";
+import R "mo:base/Result";
+import Time "mo:base/Time";
+import Timer "mo:base/Timer";
+
+import Vec "mo:vector";
+import QueueBuffer "QueueBuffer";
+
+module {
+
+  type StreamError = { #NotRegistered; #StreamClosed };
+
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   123
+  ///   func (streamId : Nat, element: Int, index: Nat): () {
+  ///     ... do your logic with incoming item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  public class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+    startFromIndex : Nat,
+  ) {
```

Comment:
We should have three more arguments to the constructor:
1. The timeout for closing the stream, a Nat in seconds.
2. A callback that is called in the error case, when first index != expected index, with first and expected index as arguments. Then the application can decide to trap or throw, or even do something else.

Reply:
Done